docs/source/jit.rst

```diff
@@ -531,13 +531,252 @@ Python-defined Constants
+        def foo(len):
+          # type: (int) -> torch.Tensor
+          rv = torch.zeros(3, 4)
+          for i in range(len):
+            if i < 10:
+                rv = rv - 1.0
+            else:
+                rv = rv + 1.0
+          return rv
+
+        print(foo.graph)
+
+    A ``ScriptModule`` with a single ``forward`` method will have an attribute
+    ``graph``, which you can use to inspect the IR representing the computation.
+    If the ScriptModule has more than one method, you will need to access
+    ``.graph`` on the method itself and not the module. We can inspect the
+    graph of a method named ``bar`` on a ScriptModule by accessing ``.bar.graph``.
+
+    The example script above produces the graph::
+
+        graph(%len : int) {
+          %13 : float = prim::Constant[value=1]()
+          %10 : int = prim::Constant[value=10]()
+          %2 : int = prim::Constant[value=4]()
+          %1 : int = prim::Constant[value=3]()
+          %3 : int[] = prim::ListConstruct(%1, %2)
+          %4 : int = prim::Constant[value=6]()
+          %5 : int = prim::Constant[value=0]()
+          %6 : int[] = prim::Constant[value=[0, -1]]()
+          %rv.1 : Dynamic = aten::zeros(%3, %4, %5, %6)
+          %8 : int = prim::Constant[value=1]()
+          %rv : Dynamic = prim::Loop(%len, %8, %rv.1)
+            block0(%i : int, %12 : Dynamic) {
+              %11 : int = aten::lt(%i, %10)
+              %rv.4 : Dynamic = prim::If(%11)
+                block0() {
+                  %14 : int = prim::Constant[value=1]()
+                  %rv.2 : Dynamic = aten::sub(%12, %13, %14)
+                  -> (%rv.2)
+                }
+                block1() {
+                  %16 : int = prim::Constant[value=1]()
+                  %rv.3 : Dynamic = aten::add(%12, %13, %16)
+                  -> (%rv.3)
+                }
+              %19 : int = prim::Constant[value=1]()
+              -> (%19, %rv.4)
+            }
+          return (%rv);
+        }
+
+    Take the instruction ``%rv.1 : Dynamic = aten::zeros(%3, %4, %5, %6)`` for
+    example. ``%rv.1 : Dynamic`` means we assign the output to a (unique)
+    value named ``rv.1``, and that value is of ``Dynamic`` type, i.e. we do
+    not know its concrete shape. ``aten::zeros`` is the operator (equivalent
+    to ``torch.zeros``) and the input list ``(%3, %4, %5, %6)`` specifies which
+    values in scope should be passed as inputs. The schema for built-in functions
+    like ``aten::zeros`` can be found at `Builtin Functions`_.
+
+    Notice that operators can also have associated ``blocks``, namely the
+    ``prim::Loop`` and ``prim::If`` operators. In the graph print-out, these
+    operators are formatted to reflect their equivalent source code forms
+    to facilitate easy debugging.
+
+    Graphs can be inspected as shown to confirm that the computation described
+    by a ``ScriptModule`` is correct, in both automated and manual fashion, as
+    described below.
```

Comment:
Isn't all this a bit too technical? We should talk more about how would you go about debugging a trace that behaves suspiciously, and the recommended workflow should definitely start with analyzing outputs of `check_trace` and not the raw IR. That should be the last resort!

Reply:
I put this here because `check_trace` often prints out graph diffs or node strings, and this as basically prerequisite information to understand what's going on there to understand the output of `check_trace`

Reply:
Can we at least say (possibly at the beginning in a note that will highlight this) that you don't need to be super familiar with this, because you hopefully won't ever need to look at those thanks to tools that we'll discuss later?

Reply:
But that isn't that case, is it? The main way the diagnostic tools present stuff is via printing IR nodes and graphs; anything else is just "some tensors don't match" or "here's a huge stack trace"

Reply:
And to be perfectly honest, I really don't think this is technical at all for the research to production crowd. These are folks who have been explicitly constructing graphs anyway; this is a clear step up

Reply:
Keep in mind that research to production crowd is not the only user group that will be interested in this functionality. Even if they know how to use the graph-based APIs, this doesn't imply that they know what SSA is, and how to understand those dumps.